### PR TITLE
Add architecture design doc

### PR DESCRIPTION
**Pull request by ShimShtein**
@ehelms @chris1984 @Ron-Lavi I would really appreciate your input about how useful this doc is to understand what's going on in the plugin, and what is missing from this doc.

### CODE REVIEW

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,223 @@
+# Architecture
+
+This plugin is intended to connect Foreman instance to Red
+Hat cloud services.
```
Comment (ehelms) [inline ARCHITECTURE.md]:
```suggestion
This plugin is intended to connect a Foreman instance to the Red
Hat Hybrid Cloud Console.
```

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+Currently it deals with the following aspects:
+
+* Inventory reports
+* Recommendations and remediations synchronization
+* Cloud side initiated remediations
+* Handling of client tools requests from the managed hosts
+
+Server side is built as a rails engine + react pages that
```
Comment (ehelms) [inline ARCHITECTURE.md]:
```suggestion
Server side is built as a Rails engine + React pages that
```

File: ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+are used for displaying information that is not related to
+existing Foreman objects.
```
Comment (ehelms) [inline ARCHITECTURE.md]: I am not sure I understand this statement.

File: ARCHITECTURE.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+In addition to these there are React components that are
+integrated into existing foreman pages, like hosts index
+and details pages by using the slot and fill mechanism.
+
+The plugin is based on asynchronous communication with the
+cloud that is done either automatically once a day or
+initiated manually from the plugin specific pages.
```
Comment (ehelms) [inline ARCHITECTURE.md]:
```suggestion
initiated manually from the plugin's UI.
```

File: ARCHITECTURE.md [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+- `{component}.scss` styling for the component
+- `{component}.test.js` tests
+
+## Inventory reports
+
+The idea is to generate a report from Foreman's inventory.
+Hosts that will be included are those that have a
+subscription attached to them. Additionally we want to
+indicate for each host whether it was reported to the cloud
+or not.
```
Comment (ehelms) [inline ARCHITECTURE.md]: Does this mean reported previously or some other method?

File: ARCHITECTURE.md [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+record that will be available to all recommendations for
+the same rule.
+
+### Foreman originated remediation
+
+Once the records were synced, there is an option to
+remediate selected recommendations by running an ansible
+playbook. Once the user selects which recommendations are
+selected for remediation, a new Remote Execution Job is
+instanciated. This REX job is based on a specific feature
```
Comment (ehelms) [inline ARCHITECTURE.md]:
```suggestion
instantiated. This REX job is based on a specific feature
```

File: ARCHITECTURE.md [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+worker and hit a new API on the Foreman server.
+The API in turn will examine the request, extract the
+playbook id from it and initiate a REX job with that id.
+The job also has a specific feature and a template. Similar
+to the Foreman initiated use case, the template calls a
+dedicated helper method, that hits a cloud API to download
+the actual playbook and then passes the playbook to the REX
+framework through the template.
+
+## Client tools requests forwarding
```
Comment (ehelms) [inline ARCHITECTURE.md:198]: I'd appreciate a bit more in this section about how the proxying happens, and some pointers to the code architecture around this. Also how authentication is handled.

File: ARCHITECTURE.md [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+separate page in the UI.
+
+The synchronization is done by the following tasks:
+1. Recommendations sync task: hits the cloud API and
+creates the recommendations record.
+2. Rules sync task: hits the cloud API and creates rule
+record that will be available to all recommendations for
+the same rule.
+
+### Foreman originated remediation
```
Comment (ehelms) [inline ARCHITECTURE.md]:
```suggestion
### Foreman initiated remediation
```

File: ARCHITECTURE.md [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+instanciated. This REX job is based on a specific feature
+and a template that comes with it. The template uses a
+single helper method that accepts the remediation ids list,
+downloads the playbook that will contain the selected
+remediations and outputs the resulting playbook. The
+downloaded playbook is passed then to the template and fed
+into the REX job. From this step it's handled by the REX
+plugin.
+
+### Cloud originated remediation (Cloud connector)
```
Comment (ehelms) [inline ARCHITECTURE.md]:
```suggestion
### Cloud initiated remediation (Cloud Connector)
```

**Comment by ehelms** [comment]
Looking great!

**Comment by ShimShtein** [comment]
@ehelms ready for another round

**Review by ehelms — APPROVED** [review]

**Review by chris1984 — APPROVED** [review]
Looks good, sorry it took so long to get to this, Fridays are good for catching up on things.

**Comment by ShimShtein** [comment]
Merging. 
Feel free to modify the doc, or ping me for more information about any of the parts.

**Pull request merged**